executor/cook/progress.py

```diff
@@ -71,37 +71,42 @@ def send_progress_update(self, progress_data, force_send=False):
+            if progress_data is not None and last_progress_data_sent != progress_data:
+                last_progress_sequence = last_progress_data_sent['progress-sequence'] if last_progress_data_sent else 0
+                if last_progress_sequence < progress_data['progress-sequence']:
+                    if force_send or self.has_enough_time_elapsed_since_last_update():
+                        logging.info('Sending progress message {}'.format(progress_data))
+                        message_dict = dict(progress_data)
+                        message_dict['task-id'] = self.task_id
+
+                        raw_progress_message = progress_data['progress-message']
+                        try:
+                            progress_str = raw_progress_message.decode('ascii').strip()
+                        except UnicodeDecodeError:
+                            logging.info('Unable to decode progress message in ascii, using empty string instead')
+                            progress_str = ''
+                        message_dict['progress-message'] = progress_str
 
-                    send_success = self.send_progress_message(progress_message)
-                    if send_success:
-                        self.last_progress_data_sent = progress_data
-                        self.last_reported_time = time.time()
+                        progress_message = json.dumps(message_dict)
+                        if len(progress_message) > self.max_message_length:
+                            num_extra_chars = len(progress_message) - self.max_message_length
+                            allowed_progress_message_length = max(len(progress_str) - num_extra_chars - 3, 0)
+                            new_progress_str = progress_str[:allowed_progress_message_length].strip() + '...'
+                            logging.info('Progress message trimmed to {}'.format(new_progress_str))
+                            message_dict['progress-message'] = new_progress_str
+                            progress_message = json.dumps(message_dict)
+
+                        send_success = self.send_progress_message(progress_message)
+                        if send_success:
+                            self.last_progress_data_sent = progress_data
+                            self.last_reported_time = time.time()
+                        else:
+                            logging.info('Unable to send progress message {}'.format(progress_message))
                     else:
-                        logging.info('Unable to send progress message {}'.format(progress_message))
+                        logging.debug('Not sending progress data as enough time has not elapsed since last update')
                 else:
-                    logging.debug('Not sending progress data as enough time has not elapsed since last update')
+                    logging.info('Skipping outdated progress data {}'.format(progress_data))
```

Comment:
I think it would be more readable if you invert the logic on these two outer branches so that the current `else` clause comes first. Since these clauses are simple (a single-line log message) and descriptive, it's helpful to see these first. As is, I end up reading over 30 lines in the true branch before hitting this one-liner in the `else` clause, at which point I have to scroll back up to see what the `if` condition was again.

Reply:
Done.